openedx/api.py

```diff
@@ -598,3 +601,65 @@ def update_edx_user_name(user):
         raise UserNameUpdateFailedException(
             "Error updating user's full name in edX.", exc
         )
+
+
+def sync_enrollments_with_edx(
+    user: User,
+) -> SyncResult[courses.models.CourseRunEnrollment]:
+    """Syncs enrollment records so that local enrollments match the enrollment data in edX"""
+    client = get_edx_api_client(user)
+    edx_enrollments = client.enrollments.get_student_enrollments()
+    local_enrollments = (
+        user.courserunenrollment_set(manager="all_objects")
+        .filter(user=user)
+        .exclude(run__courseware_id=None)
+        .order_by("run__courseware_id")
+        .all()
+    )
+    local_enrollments_map = {
+        enrollment.run.courseware_id: enrollment for enrollment in local_enrollments
+    }
+    local_only_ids, common_ids, edx_only_ids = get_partitioned_set_difference(
+        set(local_enrollments_map.keys()), set(edx_enrollments.enrollments.keys())
+    )
+    results = SyncResult()
+    # Sync active status for local enrollments that have an equivalent enrollment in edX
+    for courseware_id in common_ids:
+        edx_enrollment = edx_enrollments.enrollments[courseware_id]
+        local_enrollment = local_enrollments_map[courseware_id]
+        if local_enrollment.active and not edx_enrollment.is_active:
+            local_enrollment.deactivate_and_save(
+                change_status=ENROLL_CHANGE_STATUS_UNENROLLED
+            )
+            results.deactivated.append(local_enrollment)
+        elif not local_enrollment.active and edx_enrollment.is_active:
+            local_enrollment.reactivate_and_save()
+            results.reactivated.append(local_enrollment)
+    # Create enrollment records for any enrollments that exist in edX but not locally
+    if edx_only_ids:
+        run_values_list = courses.models.CourseRun.objects.filter(
+            courseware_id__in=edx_only_ids
+        ).values("id", "courseware_id")
+        for run_values in run_values_list:
+            local_enrollment = user.courserunenrollment_set.create(
+                user=user,
+                run_id=run_values["id"],
+                edx_enrolled=True,
+                active=edx_enrollments.enrollments[
+                    run_values["courseware_id"]
+                ].is_active,
+            )
+            results.created.append(local_enrollment)
+    # Deactivate enrollment records for any enrollments that exist locally but not in edX
+    if local_only_ids:
+        enrollments_to_deactivate = [
+            enrollment
+            for courseware_id, enrollment in local_enrollments_map.items()
+            if courseware_id in local_only_ids and enrollment.active is True
+        ]
+        for local_enrollment in enrollments_to_deactivate:
+            local_enrollment.deactivate_and_save(
```

Comment:
So one thing I think we need to watch out for here is that since this deactivates enrollments that don't show up in edx, regardless of whether they've been synced or not, meaning that it will deactivate enrollments in situations such as when there's an intermittent openedx outage. I can definitely see an enrollment failing when a learner create it, then the learner going to the dashboard and the enrollment gets deactivated. Particularly because most operations use `CourseRunEnrollments.objects` manager. I'm wondering if we actually need to delete enrollments though? I think MITx Online is supposed to be the source of truth on state, so we only really need to pull in missing enrollments to cover those that MITx Online didn't create itself, because we now have MicroMasters as a secondary source of truth.

Related to this, is the periodic task `retry_failed_edx_enrollments`, which we disabled at one point with a hard-coded return statement at the function start, but never circled back. This uses the `CourseRunEnrollments.objects` manager, so it'd be subjected to not seeing (and thus not repairing) these enrollments. I'm not proposing you fix that particular function here necessarily, although it would be a good way to test some edge cases.

Reply:
@rhysyngsun  The original reason that I thought we'd need something like this is that, at least in my local edX instance, it's possible to unenroll from a course run in the edX dashboard:

![ss 2021-09-03 at 11 38 24 ](https://user-images.githubusercontent.com/14932219/132039695-44fc93f6-10dc-457d-b13b-664f47174c34.png)

Maybe there are other ways that a user could end up being unenrolled on the edX side as well. If this happens and we don't sync, then we'd still have an enrollment record on our end that says it's active, and if they follow the link they'll see something like this (again, maybe it's different from my local instance on production):

![ss 2021-09-03 at 11 46 07 ](https://user-images.githubusercontent.com/14932219/132040572-b3863c37-0aef-4f0a-ae8d-0ec743c9b5ca.png)

If the enrollment is deactivated in edX, it seems to me that we should reflect that in our own records to avoid confusion.

Your comment did make me realize a _major_ oversight that I had a note to implement but forgot about. If that sync function fails, we should log then return local enrollments as they are. Right now the API call just fails and doesn't return anything. This function is already not going to sync anything if there's an intermittent openedx outage, but I do need to add exception handling in the enrollment API.

Let me know what you think about that scenario I described above.